### PR TITLE
Addition of test files

### PR DESCRIPTION
**Pull request by FrBier**
'Modtestlib.py' contains all functions necessary to perform certain tests on a debug module.
The individual tests are executed in the 'test_stm.py'-file.

### CODE REVIEW

**Comment by wallento** [comment]
Hi,

thanks for the test infrastructure. Generally, the pull request should go to https://github.com/opensocdebug/hardware. @imphil can assist you in how to use git subtree to push into an individual fork for that.

Another remark: Please try to keep the line width to something like 70-90. There is no strict coding convention for that, but it improves readability, for example in the code review on Github.

I will add further comments in the review.

Cheers,
Stefan

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+        # clear the buffer for the next RegRead()
+    del response[:]
+        
+    for _ in range(1):
+        yield RisingEdge(dut.clk)
+        
+    
+@cocotb.coroutine    
+def send_packet(dut, destination, source, packet_type, packet_subtype, message):
+    """function which takes necessary information as its arguments and transmits a complete packet to a chosen debug module
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+    Args:
+        dut:            device under test.
+        destination:    id of the module which is the destination.
+        source:         id of the module which is the source.
+        packet_type:    possible values are: CONST_PACKET_TYPE_REG, CONST_PACKET_TYPE_PLAIN and CONST_PACKET_TYPE_EVENT.
+        packet_subtype: choose from one of the const values: read or write requests of variable length as well as read/write responses(success/error).
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+        # determine if the last flit has to be sent and indicate this by a set 'last'-bit
+        if x == (y-1):
+            yield build_flit(dut, newmessage, 1)
+        else:
+            yield build_flit(dut, newmessage, 0)
+
+        
+@cocotb.coroutine    
+def init(dut, ID):
+    """function which initializes the testbench by setting up the clock, resetting the device under test and assigning an ID to it
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+        Returns:
+                        16 bit word containing the type of the data request for a packet.
+        
+    """
+
+    return 0x0000 | (packet_type << 14) | (packet_subtype << 10)
+ 
+@cocotb.coroutine        
+def RegWrite(dut, destination, source, reg_width, address, content, response):
+    """function which uses the subroutine 'send_packet' to write a new value into a specified register and reads in the response to tell the user if the write process was successful
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+        response:       buffer variable which the response of the dut will be stored in
+        
+        Returns:
+        -
+        
+    """
+    
+    # send a write request to the dut depending on how many data words are to be written into the memory
+    if reg_width == 16:
+        yield send_packet(dut, destination, source, CONST_PACKET_TYPE_REG, CONST_PACKET_SUBTYPE_REQ_WRITE_REG_16,  content)
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+    elif reg_width == 32:
+        yield send_packet(dut, destination, source, CONST_PACKET_TYPE_REG, CONST_PACKET_SUBTYPE_REQ_WRITE_REG_32,  content)
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+    elif reg_width == 64:
+        yield send_packet(dut, destination, source, CONST_PACKET_TYPE_REG, CONST_PACKET_SUBTYPE_REQ_WRITE_REG_64,  content)
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+    elif reg_width == 128:
+        yield send_packet(dut, destination, source, CONST_PACKET_TYPE_REG, CONST_PACKET_SUBTYPE_REQ_WRITE_REG_128, content)
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+    # if it returned an error, output that error
+    elif ((response[2] & 0x3C00) >> 10) == CONST_PACKET_SUBTYPE_REQ_WRITE_REG_ERROR:
+        print("An error occurred during the write process!")
+    # otherwise something else went wrong
+    else:
+        print("An unidentified error occurred!")
+    
+@cocotb.coroutine        
+def RegRead(dut, destination, source, reg_width, address, response):
+    """function which uses the subroutine 'send_packet' to read a value from a specified register and reads in the response to tell the user if the read process was successful
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+        # if it returned an error, output that error
+        elif ((response[2] & 0x3C00) >> 10) == CONST_PACKET_SUBTYPE_RESP_READ_REG_ERROR:
+            print("An error occurred during the read process!")
+        # otherwise something else went wrong
+        else:
+            print("An unidentified error occurred!")
+            
+    elif reg_width == 128:
+        # transmit the read request
+        yield send_packet(dut, destination, source, CONST_PACKET_TYPE_REG, CONST_PACKET_SUBTYPE_REQ_READ_REG_128,  address)
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+        # display an error message if an unsupported register width was chosen
+        raise TestFailure("Invalid register width parameter")
+    
+    for _ in range(3):
+        yield RisingEdge(dut.clk)
+
+    
+@cocotb.coroutine        
+def ModuleSubTest_STM(dut, test_id):
+    """function which takes a second parameter aside from the device under test to specify which functionality of the debug module should be tested and which tells the user if the correct values are stored inside the base registers of a debug module
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+        -
+        
+    """
+    
+    # other tests could potentially be added
+    if test_id == CONST_BASE_REG_TEST:
+            
+        response = []
+        
+        #============================================================================================================Test MOD_ID
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a break

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+        
+        # specify how the result should look like (has to be changed to defining a single value instead of a list)
+        expected_payload = []
+        expected_payload.append(1)
+        
+        #             dut, destination, source, word width, reg address, list for return value
+        yield RegRead(dut, 1,           0,      16,         0,           response)
+    
+        #                     dut, response, destination(dut), source(simulated), word width, expected payload, success variable
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Comment looks a bit odd. It is a great idea to see what the parameters mean, but probably its better to use the parameters as keyword parameters then:

    yield RegRead(dut, destination=1, ...)

For Python 3 I would suggest to use keyword only functions:

    def RegRead(*,destination,...)

This avoids errors in calling the functions.

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+        expected_payload = []
+        expected_payload.append(4)
+        
+        #             dut, destination, source, word width, reg address, list for return value
+        yield RegRead(dut, 1,           0,      16,         1,           response)
+    
+        #                     dut, response, destination(dut), source(simulated), word width, expected payload, success variable
+        yield evaluate_packet(dut, response, 1,                0,                 16,         expected_payload)
+                
+        #========================================================================================================Test MOD_VENDOR
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a shorter width

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+        expected_payload = []
+        expected_payload.append(0)
+        
+        #             dut, destination, source, word width, reg address, list for return value
+        yield RegRead(dut, 1,           0,      16,         2,           response)
+    
+        #                     dut, response, destination(dut), source(simulated), word width, expected payload, success variable
+        yield evaluate_packet(dut, response, 1,                0,                 16,         expected_payload)
+                
+        #============================================================================================================Test MOD_CS
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a shorter width

File: external/opensocdebug/hardware/modules/stm/modtestlib.py [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+        # expected_payload = []
+        # expected_payload.append(1)
+        #
+        #               dut, destination, source, word width, reg address, list for return value
+        # yield RegRead(dut, 1,           0,      16,         3,           response)
+        #
+        #                       dut, response, destination(dut), source(simulated), word width, expected payload, success variable
+        # yield evaluate_packet(dut, response, 1,                0,                 16,         expected_payload)
+        #
+        #====================================================================================================Test MOD_EVENT_DEST
```
Comment (wallento) [inline external/opensocdebug/hardware/modules/stm/modtestlib.py]: Please improve readability with a shorter width

**Comment by imphil** [comment]
It's a wip commit which should be merged to a devel branch for now. Once it all works well split it into the subprojects. I'll take care of that.

**Comment by wallento** [comment]
Okay, I see. Please have a look at the comments, especially python 3 keyword only, which I find very useful. Cheers

**Comment by imphil** [comment]
Thanks Stefan for your comments. @FrBier please have a look at those and let's move the further discussion to https://github.com/opensocdebug/hardware/issues/35 as outlined there.

**Pull request closed without merging**